logstash-core/lib/logstash/instrument/metric.rb

```diff
@@ -33,19 +56,23 @@ def gauge(key, value)
       collector.push(namespace_information, key, :gauge, :set, value)
     end
 
-    def time(key, &block)
+    def time(key)
       validate_key!(key)
+
       if block_given?
-        start_time = Time.now
-        content = block.call
-        duration = Time.now - start_time
-        gauge(key, duration)
+        timer = TimedExecution.new(self, key)
+        content = yield
+        timer.stop
         return content
       else
-        raise MetricNoBlockProvided
+        TimedExecution.new(self, key)
```

Comment:
Something you don't want to capture a block, this method support both kind of invocation.

```ruby
timer = metric.time("responsetime")
# do work
timer.stop

or 

metric.time("responsetime")  do
# do work
end
```